src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,807 @@
+	// fashion if we write our own decoder or expose the one in the underlying library.
+	if err := enc.unmarshaled.Unmarshal(ant); err != nil {
+		return fmt.Errorf(
+			"%s error unmarshaling annotation into proto message: %v", encErrPrefix, err)
+	}
+
+	if enc.numEncoded == 0 {
+		enc.encodeHeader()
+	}
+
+	// Control bit that indicates the stream has more data.
+	enc.stream.WriteBit(opCodeMoreData)
+
+	if err := enc.encodeTimestamp(dp.Timestamp, tu); err != nil {
+		return fmt.Errorf(
+			"%s error encoding timestamp: %v", encErrPrefix, err)
+	}
+
+	enc.encodeProto(enc.unmarshaled)
+	enc.numEncoded++
+	enc.lastEncodedDP = dp
+	return nil
+}
+
+// Stream returns a copy of the underlying data stream.
+func (enc *Encoder) Stream() xio.SegmentReader {
+	seg := enc.segment(true)
+	if seg.Len() == 0 {
+		return nil
+	}
+
+	if readerPool := enc.opts.SegmentReaderPool(); readerPool != nil {
+		reader := readerPool.Get()
+		reader.Reset(seg)
+		return reader
+	}
+	return xio.NewSegmentReader(seg)
+}
+
+func (enc *Encoder) segment(copy bool) ts.Segment {
+	length := enc.stream.Len()
+	if enc.stream.Len() == 0 {
+		return ts.Segment{}
+	}
+
+	// TODO: Should this be cropping out the last byte?
+	var head checked.Bytes
+	buffer, _ := enc.stream.Rawbytes()
+	if !copy {
+		// Take ref from the ostream.
+		head = enc.stream.Discard()
+	} else {
+		// Copy into new buffer.
+		head = enc.newBuffer(length)
+		head.IncRef()
+		head.AppendAll(buffer)
+		head.DecRef()
+	}
+
+	return ts.NewSegment(head, nil, ts.FinalizeHead)
+}
+
+// NumEncoded returns the number of encoded messages.
+func (enc *Encoder) NumEncoded() int {
+	return enc.numEncoded
+}
+
+// LastEncoded returns the last encoded datapoint. Does not include
+// annotation / protobuf message for interface purposes.
+func (enc *Encoder) LastEncoded() (ts.Datapoint, error) {
+	if enc.numEncoded == 0 {
+		return ts.Datapoint{}, errNoEncodedDatapoints
+	}
+
+	return enc.lastEncodedDP, nil
+}
+
+// Len returns the length of the data stream.
+func (enc *Encoder) Len() int {
+	return enc.stream.Len()
+}
+
+func (enc *Encoder) encodeHeader() {
+	enc.encodeVarInt(currentEncodingSchemeVersion)
+	enc.encodeVarInt(uint64(enc.opts.ByteFieldDictionaryLRUSize()))
+	enc.encodeCustomSchemaTypes()
+}
+
+func (enc *Encoder) encodeCustomSchemaTypes() {
+	if len(enc.customFields) == 0 {
+		enc.encodeVarInt(0)
+		return
+	}
+
+	// Field numbers are 1-indexed so encoding the maximum field number
+	// at the beginning is equivalent to encoding the number of types
+	// we need to read after if we imagine that we're encoding a 1-indexed
+	// bitset where the position in the bitset encodes the field number (I.E
+	// the first value is the type for field number 1) and the values are
+	// the number of bits required to unique identify a custom type instead of
+	// just being a single bit (3 bits in the case of version 1 of the encoding
+	// scheme.)
+	maxFieldNum := enc.customFields[len(enc.customFields)-1].fieldNum
```

Comment:
correct although I can add a sort just to be extra safe

Reply:
I mean if it's in the proto spec then no worries, but otherwise it might be better for us to be safe in case they change that later for whatever reason.